proof-systems/src/darlin/poly-commit/src/constraints/tests.rs

```diff
@@ -0,0 +1,584 @@
+
+        let point = G::ScalarField::rand(rng);
+        let mut value = polynomial.evaluate(point);
+        let fs_seed = String::from("TEST_SEED").into_bytes();
+        let mut fs_rng =
+            PC::RandomOracle::from_seed(fs_seed.clone()).map_err(|err| PolyError::from(err))?;
+        let proof = PC::open(
+            &ck,
+            polynomial,
+            point,
+            is_hiding,
+            randomness,
+            &mut fs_rng,
+            Some(rng),
+        )?;
+        // check that verify with primitive succeeds
+        let mut fs_rng =
+            PC::RandomOracle::from_seed(fs_seed.clone()).map_err(|err| PolyError::from(err))?;
+        let v_state = PC::succinct_verify(&vk, &commitment, point, value, &proof, &mut fs_rng)?;
+        if v_state.is_none() {
+            Err(PolyError::FailedSuccinctCheck)?
+        }
+
+        let mut cs = ConstraintSystem::<ConstraintF>::new(SynthesisMode::Debug);
+        match test_conf.negative_type {
+            Some(NegativeTestType::Commitments) => {
+                commitment = commitment * &G::ScalarField::rand(rng)
+            }
+            Some(NegativeTestType::Values) => value = G::ScalarField::rand(rng),
+            None => {}
+        }
+        let vk_gadget = PCG::VerifierKey::alloc(cs.ns(|| "alloc verifier key"), || Ok(vk))?;
+        let commitment_gadget = PCG::Commitment::alloc(cs.ns(|| "alloc commitment"), || {
+            Ok(commitment_for_alloc::<G, PC>(
+                &commitment,
+                &test_conf.negative_type,
+            ))
+        })?;
+
+        let point_gadget = NonNativeFieldGadget::<G::ScalarField, ConstraintF>::alloc(
+            cs.ns(|| "alloc evaluation point"),
+            || Ok(point),
+        )?;
+        let value_gadget = Boolean::alloc_input_vec(
+            cs.ns(|| "alloc polynomial evalauation on point"),
+            value.write_bits().as_slice(),
+        )?;
+        let proof_gadget = PCG::Proof::alloc(cs.ns(|| "alloc opening proof"), || Ok(proof))?;
+        let mut fs_gadget = PCG::RandomOracle::init_from_seed(cs.ns(|| "init fs oracle"), fs_seed)?;
+        let _v_state_gadget = PCG::succinct_verify(
+            cs.ns(|| "succinct-verify"),
+            &vk_gadget,
+            &commitment_gadget,
+            &point_gadget,
+            &value_gadget,
+            &proof_gadget,
+            &mut fs_gadget,
+        )?;
+        let successful_test = cs.is_satisfied() ^ test_conf.negative_type.is_some();
+        if !successful_test {
+            println!("{:?}", cs.which_is_unsatisfied());
+        }
+        assert!(successful_test);
+
+        // test mul_bits_fixed_base for commitments
+        let bits = fs_gadget.enforce_get_challenge::<_, 128>(cs.ns(|| "get random bits for mul_bits_fixed_base"))?;
+        PCG::Commitment::mul_bits_fixed_base(&commitment, cs, &bits[..])?;
+    }
+    Ok(())
+}
+
+fn test_multi_point_multi_poly_verify<
```

Comment:
Would be nice to have a test template, much like the primitive, to reduce a bit code duplication, at least for what concerns proof generation before calling the specific verification functions